plugins/lua/buildingplan/unlink_mechanisms.lua

```diff
@@ -0,0 +1,680 @@
+    self.subviews.scroll:update(scroll_pos + 1, self.num_buttons*3, #self.links*3)
+    self:scroll(0)
+
+    local bci_len = #self.building.contained_items
+    for i=1, self.num_buttons do
+        local button = self:get_button(i, true)
+        local offset = i*3 - 1 - ((scroll_pos + 1) % 3)
+        local idx = self:idx_from_offset(offset)
+
+        button.visible = false
+        if idx > 0 and idx < bci_len then
+            button.frame.t = offset
+            button.visible = true
+        end
+        button:updateLayout()
+    end
+
+    local b = (self.frame.h % 3) == 1 and #self.links >= self.num_buttons and 0 or 1
+    self.subviews.unlink_mode.frame.b = b --avoid overlapping list
+    self.subviews.unlink_all.frame.b = b
+    self.subviews.unlink_mode:updateLayout()
+    self.subviews.unlink_all:updateLayout()
+end
+
+function MechLinkOverlay:preUpdateLayout(parent_rect)
+    for i=1, self.num_buttons do --hide existing buttons
+        local button = self:get_button(i)
+        if button then
+            button.visible = false
+        end
+    end
+
+    local h = parent_rect.height - 49
+    self.frame.h = h + 1 --includes lower border
+    self.num_buttons = h // 3
+
+    self.subviews.scroll.frame.h = self.num_buttons*3
+end
+
+function MechLinkOverlay:onRenderFrame(dc, rect)
+    if self.bld_id ~= sheet.viewing_bldid then
+        self.bld_id = sheet.viewing_bldid
+        self.building = df.building.find(self.bld_id)
+    end
+
+    self:update_buttons()
+
+    MechLinkOverlay.super.onRenderFrame(self, dc, rect)
+end
+
+-- ----------------------
+-- MechItemOverlay
+--
+
+MechItemOverlay = defclass(MechItemOverlay, overlay.OverlayWidget)
+MechItemOverlay.ATTRS
+{
+    desc = "Allows freeing unlinked mechanisms from buildings.",
+    default_enabled = true,
+    overlay_only = true,
+    default_pos = {x=-41, y=-4},
+    frame = {w=56, h=27},
+    viewscreens = {},
+}
+
+for _,v in ipairs(valid_build) do
+    table.insert(MechItemOverlay.ATTRS.viewscreens, "dwarfmode/ViewSheets/BUILDING/"..v.."/Items")
+end
+valid_build = nil
+
+function MechItemOverlay:init()
+    self.num_buttons = 0
+
+    self:addviews
+    {
+        widgets.HotkeyLabel
+        {
+            view_id = "free_all",
+            frame = {b=1, r=3},
+            key = "CUSTOM_SHIFT_F",
+            label = "Free all",
+            auto_width = true,
+            on_activate = self:callback("ask_free_all"),
+        },
+    }
+end
+
+function MechItemOverlay:idx_from_offset(offset)
+    if offset < 0 or offset >= self.num_buttons*3 then
+        return 0 --outside of list
+    else
+        return (offset + sheet.scroll_position_item) // 3
+    end
+end
+
+function MechItemOverlay:get_button(n, ensure)
+    local button = self.subviews["free_"..n]
+    if not button and ensure then
+        self:addviews
+        {
+            widgets.TextButton
+            {
+                view_id = "free_"..n,
+                frame = {t=0, l=28, w=6, h=1},
```

Comment:
I'd move this to the right two or three tiles to give longer names a little more room:
![image](https://github.com/DFHack/dfhack/assets/977482/69e687b7-12b6-4215-b930-dcbd39d955a8)

Reply:
There can be an extra button on that spot, IIRC.

Reply:
![free](https://github.com/DFHack/dfhack/assets/25622708/e3280fb8-c03e-4f0b-bfab-0e9e62995b11)
:(